hushline/premium.py

```diff
@@ -0,0 +1,536 @@
+            subscription.current_period_end
+        )
+        user.stripe_subscription_current_period_start = datetime.fromtimestamp(
+            subscription.current_period_start
+        )
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with customer ID {subscription.customer}")
+
+
+def handle_subscription_updated(subscription: stripe.Subscription) -> None:
+    # customer.subscription.updated
+
+    # If subscription changes to cancel or unpaid, downgrade user
+    user = db.session.query(User).filter_by(stripe_subscription_id=subscription.id).first()
+    if user:
+        user.stripe_subscription_status = StripeSubscriptionStatusEnum(subscription.status)
+        user.stripe_subscription_cancel_at_period_end = subscription.cancel_at_period_end
+        user.stripe_subscription_current_period_end = datetime.fromtimestamp(
+            subscription.current_period_end
+        )
+        user.stripe_subscription_current_period_start = datetime.fromtimestamp(
+            subscription.current_period_start
+        )
+
+        current_app.logger.info("status is: " + subscription.status)
+        if subscription.status in ["active", "trialing"]:
+            user.tier_id = BUSINESS_TIER
+        else:
+            user.tier_id = FREE_TIER
+
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with subscription ID {subscription.id}")
+
+
+def handle_subscription_deleted(subscription: stripe.Subscription) -> None:
+    # customer.subscription.deleted
+
+    user = db.session.query(User).filter_by(stripe_subscription_id=subscription.id).first()
+    if user:
+        user.tier_id = FREE_TIER
+        user.stripe_subscription_id = None
+        user.stripe_subscription_status = None
+        user.stripe_subscription_cancel_at_period_end = None
+        user.stripe_subscription_current_period_end = None
+        user.stripe_subscription_current_period_start = None
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find user with subscription ID {subscription.id}")
+
+
+def handle_invoice_created(invoice: stripe.Invoice) -> None:
+    # invoice.created
+
+    try:
+        new_invoice = StripeInvoice(invoice)
+        db.session.add(new_invoice)
+        db.session.commit()
+    except ValueError as e:
+        current_app.logger.error(f"Error creating invoice: {e}")
+
+
+def handle_invoice_updated(invoice: stripe.Invoice) -> None:
+    # invoice.updated
+
+    stripe_invoice = db.session.query(StripeInvoice).filter_by(invoice_id=invoice.id).first()
+    if stripe_invoice:
+        stripe_invoice.total = invoice.total
+        stripe_invoice.status = StripeInvoiceStatusEnum(invoice.status)
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find invoice with ID {invoice.id}")
+
+
+async def worker(app: Flask) -> None:
+    current_app.logger.error("Starting worker")
+    with app.app_context():
+        while True:
```

Comment:
Maybe not for this PR but as followup, do we want to set up some kinda signal handler / semaphore here instead of `while True` just so we can avoid possible errant state if this is killed in the middle of processing?

Reply:
Good idea for the future. I'm not sure the best way to handle it. Though I expect we will only actually want one worker container running at any time, since it's only used for Stripe webhook events. Hush Line would have to be getting a massive amounts of money for this to hit any scaling problems